cip/CIP2015-11-09-type-coercions.adoc

```diff
@@ -0,0 +1,98 @@
+= CIP2015-11-09 - Type coercions
+:numbered:
+:toc:
+:toc-placement: macro
+:source-highlighter: codemirror
+
+*Author:* Henrik Nyman henrik.nyman@neotechnology.com, Mats Rydberg mats@neotechnology.com
+
+
+[abstract]
+.Abstract
+--
+This is a retroactive CIP aiming to describe the current type coercions that exist in today's Cypher.
+--
+
+toc::[]
+
+
+== Motivation
+
+Due to recent issues with (the lack of) type coercions, the need for a requirements specification has become actualized.
+
+== Background
+
+In the case of a collection with both integer and floating-point numbers, it's type will be a collection of numbers.
+Functions that work on both integers and floats would then accept all elements of this collection one-by-one, but if they do not also accept type number they will not be able to act on the collection itself when used in a comprehension.
+In this type of situation, it is very valuable for a user to know what kinds of coercions that Cypher will perform, and what explicit casting that they will have to do themselves.
+
+== Proposal
+
+The current Cypher behavior is proposed, which consists of two type coercions: `CTCollection` to `CTBoolean`, and `CTInteger` to `CTFloat`.
+
+=== Examples
+
+Here follows a few queries in which types are coerced.
+
+Check if a collection is empty, and if it is not, return the first element:
+[source, cypher]
+----
+WITH [] as coll
+WHERE coll
+RETURN coll[0]
+----
+
+Return the size of a non-empty collection:
+[source, cypher]
+----
+WITH [1,2,3.5] as coll
+WHERE coll
+RETURN size(coll)
+----
+
+Calculate the cosine of a value:
+[source, cypher]
+----
+WITH 1 as int
+RETURN cos(int)
+----
+
+=== Syntax
+
+No syntax changes are proposed.
+
+=== Semantics
+
+There are no new semantics to consider.
+
+=== Alternatives
+
+There are plenty of additional coercions to consider.
```

Comment:
We could limit int to float coercion for literals only; that perhaps is less surprising and addressed the 80/20 of why that coercion is practical.